diff --git a/extension/json/CMakeLists.txt b/extension/json/CMakeLists.txt
index 4101111df8..3e035e80d3 100644
--- a/extension/json/CMakeLists.txt
+++ b/extension/json/CMakeLists.txt
@@ -34,6 +34,7 @@ set(JSON_EXTENSION_FILES
 build_static_extension(json ${JSON_EXTENSION_FILES})
 set(PARAMETERS "-warnings")
 build_loadable_extension(json ${PARAMETERS} ${JSON_EXTENSION_FILES})
+target_link_libraries(json_loadable_extension duckdb_yyjson)
 
 install(
   TARGETS json_extension
