diff --git a/src/include/duckdb/common/hive_partitioning.hpp b/src/include/duckdb/common/hive_partitioning.hpp
index 055dc496c5..ce14176dae 100644
--- a/src/include/duckdb/common/hive_partitioning.hpp
+++ b/src/include/duckdb/common/hive_partitioning.hpp
@@ -14,7 +14,7 @@
 #include "duckdb/optimizer/statistics_propagator.hpp"
 #include "duckdb/planner/expression_iterator.hpp"
 #include "duckdb/planner/table_filter.hpp"
-#include "re2/re2.h"
+#include "../../third_party/re2/re2/re2.h"
 
 #include <iostream>
 #include <sstream>
diff --git a/third_party/re2/re2/re2.h b/third_party/re2/re2/re2.h
index 56a664fc91..7e92caf9ab 100644
--- a/third_party/re2/re2/re2.h
+++ b/third_party/re2/re2/re2.h
@@ -196,7 +196,7 @@
 #include <mutex>
 #include <string>
 
-#include "re2/stringpiece.h"
+#include "stringpiece.h"
 
 namespace duckdb_re2 {
 class Prog;
