diff --git a/src/common/arrow/arrow_appender.cpp b/src/common/arrow/arrow_appender.cpp
index 83f190b570..60c4b36268 100644
--- a/src/common/arrow/arrow_appender.cpp
+++ b/src/common/arrow/arrow_appender.cpp
@@ -40,8 +40,10 @@ void ArrowAppender::Append(DataChunk &input, const idx_t from, const idx_t to, c
 	for (idx_t i = 0; i < input.ColumnCount(); i++) {
 		if (root_data[i]->extension_data && root_data[i]->extension_data->duckdb_to_arrow) {
 			Vector input_data(root_data[i]->extension_data->GetInternalType());
+			if (options.client_context) {
 			root_data[i]->extension_data->duckdb_to_arrow(*options.client_context, input.data[i], input_data,
 			                                              input_size);
+			}
 			root_data[i]->append_vector(*root_data[i], input_data, from, to, input_size);
 		} else {
 			root_data[i]->append_vector(*root_data[i], input.data[i], from, to, input_size);
