diff --git a/src/planner/binder/statement/bind_copy.cpp b/src/planner/binder/statement/bind_copy.cpp
index ce2f93ad86..140af5ba36 100644
--- a/src/planner/binder/statement/bind_copy.cpp
+++ b/src/planner/binder/statement/bind_copy.cpp
@@ -152,7 +152,9 @@ BoundStatement Binder::BindCopyTo(CopyStatement &stmt, CopyToType copy_to_type)
 		}
 	}
 	bool is_remote_file = FileSystem::IsRemoteFile(stmt.info->file_path);
-	if (is_remote_file) {
+	if ( is_remote_file ) {
+		use_tmp_file = false;
+	} else if( context.db->config.options.use_direct_io ) {
 		use_tmp_file = false;
 	} else {
 		auto &fs = FileSystem::GetFileSystem(context);
