diff --git a/src/storage/storage_manager.cpp b/src/storage/storage_manager.cpp
index 456cd22614..c4777dd28e 100644
--- a/src/storage/storage_manager.cpp
+++ b/src/storage/storage_manager.cpp
@@ -163,9 +163,12 @@ void SingleFileStorageManager::LoadDatabase(const optional_idx block_alloc_size)
 	options.use_direct_io = config.options.use_direct_io;
 	options.debug_initialize = config.options.debug_initialize;

-	// Check if the database file already exists.
-	// Note: a file can also exist if there was a ROLLBACK on a previous transaction creating that file.
-	if (!read_only && !fs.FileExists(path)) {
+    auto db_file_handle = fs.OpenFile(path, FileFlags::FILE_FLAGS_READ | FileFlags::FILE_FLAGS_NULL_IF_NOT_EXISTS);
+    bool is_empty_file = db_file_handle->GetFileSize() == 0;
+    db_file_handle.reset();
+
+ 	// first check if the database exists
+	if (!read_only && ( !fs.FileExists(path) || ( options.use_direct_io && is_empty_file )) ) {
 		// file does not exist and we are in read-write mode
 		// create a new file
 
