diff --git a/src/storage/storage_manager.cpp b/src/storage/storage_manager.cpp
index 45d42174f1..10b3ce47b1 100644
--- a/src/storage/storage_manager.cpp
+++ b/src/storage/storage_manager.cpp
@@ -162,9 +162,15 @@ void SingleFileStorageManager::LoadDatabase(StorageOptions storage_options) {
 			    row_group_size, STANDARD_VECTOR_SIZE);
 		}
 	}
-	// Check if the database file already exists.
-	// Note: a file can also exist if there was a ROLLBACK on a previous transaction creating that file.
-	if (!read_only && !fs.FileExists(path)) {
+	bool is_empty_file = true;
+	auto db_file_handle = fs.OpenFile(path, FileFlags::FILE_FLAGS_READ | FileFlags::FILE_FLAGS_NULL_IF_NOT_EXISTS);
+	if (db_file_handle && db_file_handle->GetFileSize() != 0) {
+		is_empty_file = false;
+		db_file_handle.reset();
+	}
+
+	// first check if the database exists
+	if (!read_only && (!fs.FileExists(path) || (options.use_direct_io && is_empty_file))) {
 		// file does not exist and we are in read-write mode
 		// create a new file
 
