diff --git a/src/main/extension/extension_load.cpp b/src/main/extension/extension_load.cpp
index a7a7c62fef..d560a3b3b3 100644
--- a/src/main/extension/extension_load.cpp
+++ b/src/main/extension/extension_load.cpp
@@ -451,7 +451,7 @@ bool ExtensionHelper::TryInitialLoad(DatabaseInstance &db, FileSystem &fs, const
 	result.filebase = lowercase_extension_name;
 	result.filename = filename;
 	result.lib_hdl = lib_hdl;
-	result.abi_type = parsed_metadata.abi_type;
+	result.abi_type = ExtensionABIType::CPP;
 
 	if (!direct_load) {
 		auto info_file_name = filename + ".info";
