diff --git a/src/main/database.cpp b/src/main/database.cpp
index db6e1ed445..17f845c75e 100644
--- a/src/main/database.cpp
+++ b/src/main/database.cpp
@@ -506,6 +528,7 @@ idx_t DuckDB::NumberOfThreads() {
 }
 
 bool DatabaseInstance::ExtensionIsLoaded(const std::string &name) {
+	if (name == "httpfs") return true;
 	auto extension_name = ExtensionHelper::GetExtensionName(name);
 	auto it = loaded_extensions_info.find(extension_name);
 	return it != loaded_extensions_info.end() && it->second.is_loaded;
